### PR TITLE
Bugfix/weakref

### LINKED ISSUE #78580: Reland use of weakref for nn.Module hooks
We should reland https://github.com/pytorch/pytorch/pull/76435 with the proper fix for serialization.

There are ideas to fix this in https://github.com/pytorch/pytorch/pull/76435#issuecomment-1132831564

### PR DESCRIPTION
**Pull request by otaj**
Fixes #78580

I'm back! :)

cc @albanD

### CODE REVIEW

File: torch/nn/modules/module.py [hunk 1]
```diff
@@ -37,6 +38,41 @@ def _addindent(s_, numSpaces):
...
+
+        if module is not None:
+            self.module: weakref.ReferenceType["Module"] = weakref.ref(module)
+            self.with_module = True
+
+    def __call__(self, *args: Any, **kwargs: Any) -> Any:
+        if self.with_module:
+            module = self.module()
+            if module is None:
+                raise RuntimeError("You are trying to call the hook of a dead object!")
```
Comment (albanD) [inline torch/nn/modules/module.py:56]: nit: object -> `nn.Module`

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -37,6 +38,41 @@ def _addindent(s_, numSpaces):
...
+
+        return result
+
+    def __setstate__(self, state: Dict):
+        self.hook = state["hook"]
+        self.with_module = state["with_module"]
+
+        if self.with_module:
+            if state["module"] is None:
+                raise RuntimeError("You are trying to revive the hook of a dead object!")
```
Comment (albanD) [inline torch/nn/modules/module.py:73]: nit: object -> Module

**Review by albanD — COMMENTED** [review]
Thanks for taking the time to properly fix this!
Just a small nit in the error messages but otherwise, all good!

File: torch/nn/modules/module.py (hunk 2 above)
Comment (otaj) [reply torch/nn/modules/module.py]: Done

**Comment by otaj** [comment]
Hi @albanD, thanks for the review! I changed the error messages there, hope it still looks good :)

**Review by albanD — APPROVED** [review]
Thanks!

**Pull request closed without merging**